controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/HealthCheckTask.java

```diff
@@ -0,0 +1,166 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.pinterest.rocksplicator.controller.tasks;
+
+import com.pinterest.rocksplicator.controller.bean.ClusterBean;
+import com.pinterest.rocksplicator.controller.bean.HostBean;
+import com.pinterest.rocksplicator.controller.bean.SegmentBean;
+import com.pinterest.rocksplicator.controller.bean.ShardBean;
+import com.pinterest.rocksplicator.controller.config.ConfigParser;
+import com.pinterest.rocksplicator.controller.util.AdminClientFactory;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+
+
+/**
+ * @author Ang Xu (angxu@pinterest.com)
+ */
+public class HealthCheckTask extends TaskBase<HealthCheckTask.Param> {
+
+  public static final Logger LOG = LoggerFactory.getLogger(HealthCheckTask.class);
+
+  @Inject
+  public CuratorFramework zkClient;
+
+  @Inject
+  public AdminClientFactory clientFactory;
+
+  public HealthCheckTask(Param param) {
+    super(param);
+  }
+
+  @Override
+  public void process(Context ctx) throws Exception {
+    final String clusterName = getParameter().getClusterName();
+    final String zkPath = getParameter().getZkPath();
+    final InterProcessLock lock =
+        new InterProcessMutex(zkClient,  zkPath + clusterName + ".lock");
+    lock.acquire();
+    try {
+      byte[] data = zkClient.getData().forPath(zkPath + clusterName);
+      ClusterBean clusterBean = ConfigParser.parseClusterConfig(clusterName, data);
+
+      Set<InetSocketAddress> hosts = new HashSet<>();
+      for (SegmentBean segmentBean : clusterBean.getSegments()) {
+        for (HostBean hostBean : segmentBean.getHosts()) {
+          hosts.add(new InetSocketAddress(hostBean.getIp(), hostBean.getPort()));
+        }
+        // check segment replica
+        checkSegment(segmentBean, getParameter().getNumReplicas());
+      }
+
+      // ping all hosts
+      for (InetSocketAddress hostAddr : hosts) {
+        clientFactory.getClient(hostAddr).ping();
+      }
+
+      LOG.info("All hosts are good");
+
+      ctx.getTaskQueue().finishTask(ctx.getId(),
+          String.format("Cluster %s is healthy", clusterName));
+    } catch (Exception ex) {
+      ctx.getTaskQueue().failTask(ctx.getId(),
+          String.format("Cluster %s is unhealthy, reason = %s", clusterName, ex.getMessage()));
```

Comment:
we will be interested to know which ip is not pingable. do we have ip in the exception?